org.eclipse.scanning.test/src/org/eclipse/scanning/test/command/AbstractMScanTest.java

```diff
@@ -0,0 +1,432 @@
+			subscriber.addListener(new IScanListener() {
+				@Override
+				public void scanEventPerformed(ScanEvent evt) {
+					if (evt.getBean().getStatus()==Status.FAILED) failed.add(evt.getBean());
+					if (evt.getBean().getPosition()!=null) {
+						beans.add(evt.getBean());
+					}
+				}
+	
+				@Override
+				public void scanStateChanged(ScanEvent evt) {
+					if (evt.getBean().getStatus()==Status.FAILED) failed.add(evt.getBean());
+					if (evt.getBean().scanStart()) {
+						startEvents.add(evt.getBean()); // Should be just one
+					}
+	                if (evt.getBean().scanEnd()) {
+	                	latch.countDown();
+	                }
+				}
+			});
+	
+			
+			// Ok done that, now we sent it off...
+			pi.exec("submit("+name+", block="+(blocking?"True":"False")+", broker_uri='"+uri+"')");
+			
+			Thread.sleep(200);
+			boolean ok = latch.await(maxScanTimeS, TimeUnit.SECONDS);
+			if (!ok) throw new Exception("The latch broke before the scan finished!");
+			
+			if (failed.size()>0) throw new Exception(failed.get(0).getMessage());
+			
+			ScanBean start = startEvents.get(0);
+			assertEquals(start.getSize(), beans.size());
+			assertEquals(1, startEvents.size());
+			
+			return beans;
+			
+		} finally {
+			subscriber.disconnect();
+			submitter.disconnect();
+		}
+	}
+
+	protected List<ScanBean> runAndCheck(String name, String mainDetectorName, String processingDetectorName, boolean blocking, long maxScanTimeS) throws Exception {
+		
+		final IEventService eservice = Services.getEventService();
+
+		// Let's listen to the scan and see if things happen when we run it
+		final ISubscriber<IScanListener> subscriber = eservice.createSubscriber(new URI(servlet.getBroker()), servlet.getStatusTopic());
+		final ISubmitter<ScanBean>       submitter  = eservice.createSubmitter(new URI(servlet.getBroker()),  servlet.getSubmitQueue());
+		
+		try {
+			final List<ScanBean> beans = new ArrayList<>(13);
+			final List<ScanBean> failed = new ArrayList<>(13);
+			final List<ScanBean> startEvents = new ArrayList<>(13);
+			
+			final CountDownLatch latch = new CountDownLatch(1);
+			subscriber.addListener(new IScanListener() {
+				@Override
+				public void scanEventPerformed(ScanEvent evt) {
+					if (evt.getBean().getStatus()==Status.FAILED) failed.add(evt.getBean());
+					if (evt.getBean().getPosition()!=null) {
+						beans.add(evt.getBean());
+					}
+				}
+	
+				@Override
+				public void scanStateChanged(ScanEvent evt) {
+					if (evt.getBean().getStatus()==Status.FAILED) failed.add(evt.getBean());
+					if (evt.getBean().scanStart()) {
+						startEvents.add(evt.getBean()); // Should be just one
+					}
+	                if (evt.getBean().scanEnd()) {
+	                	latch.countDown();
+	                }
+				}
+			});
+	
+			
+			// Ok done that, now we sent it off...
+			pi.exec("submit("+name+", block="+(blocking?"True":"False")+", broker_uri='"+uri+"')");
+			
+			Thread.sleep(200);

```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Remove this use of "Thread.sleep()". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS2925)